File: src/packages/transactions/envelope.c
#include "envelope.h"

#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <openssl/sha.h>

#include "packages/keystore/keystore.h"
#include "packages/signing/signing.h"
#include "packages/utils/logger.h"

#include "envelope.pb-c.h"

static int compute_envelope_digest(const Tinyweb__EnvelopeHeader* hdr,
                                   const unsigned char* cipher,
                                   size_t cipher_len,
                                   unsigned char out_hash[SHA256_DIGEST_LENGTH]) {
    if (!hdr || !cipher) return -1;

    // Hash ciphertext first
    unsigned char payload_hash[SHA256_DIGEST_LENGTH];
    SHA256(cipher, cipher_len, payload_hash);

    // Compute digest from raw header data (not protobuf-serialized)
    // This matches the frontend approach and is cleaner
    SHA256_CTX ctx;
    SHA256_Init(&ctx);
    
    // Domain separator
    static const unsigned char domain[] = { 'T','W','E','N','V','E','L','O','P','E','\0' };
    SHA256_Update(&ctx, domain, sizeof(domain));
    
    // Header fields in canonical order:
    // 1. version (uint32, 4 bytes, little-endian)
    uint32_t version = hdr->version;
    SHA256_Update(&ctx, (unsigned char*)&version, sizeof(version));
    
    // 2. content_type (uint32, 4 bytes, little-endian)
    uint32_t content_type = hdr->content_type;
    SHA256_Update(&ctx, (unsigned char*)&content_type, sizeof(content_type));
    
    // 3. schema_version (uint32, 4 bytes, little-endian)
    uint32_t schema_version = hdr->schema_version;
    SHA256_Update(&ctx, (unsigned char*)&schema_version, sizeof(schema_version));
    
    // 4. timestamp (uint64, 8 bytes, little-endian)
    uint64_t timestamp = hdr->timestamp;
    SHA256_Update(&ctx, (unsigned char*)&timestamp, sizeof(timestamp));
    
    // 5. sender_pubkey (32 bytes)
    if (hdr->sender_pubkey.len != 32) return -1;
    SHA256_Update(&ctx, hdr->sender_pubkey.data, 32);
    
    // 6. recipients_pubkey (count + data for each)
    uint32_t num_recipients = hdr->n_recipients_pubkey;
    SHA256_Update(&ctx, (unsigned char*)&num_recipients, sizeof(num_recipients));
    for (size_t i = 0; i < num_recipients; i++) {
        if (hdr->recipients_pubkey[i].len != 32) return -1;
        SHA256_Update(&ctx, hdr->recipients_pubkey[i].data, 32);
    }
    
    // 7. group_id (length + data, or empty)
    uint32_t group_id_len = hdr->group_id.len;
    SHA256_Update(&ctx, (unsigned char*)&group_id_len, sizeof(group_id_len));
    if (group_id_len > 0) {
        SHA256_Update(&ctx, hdr->group_id.data, group_id_len);
    }
    
    // 8. Payload hash
    SHA256_Update(&ctx, payload_hash, sizeof(payload_hash));
    
    SHA256_Final(out_hash, &ctx);
    return 0;
}

int tw_envelope_serialize(const Tinyweb__Envelope* env, unsigned char** out, size_t* out_len) {
    if (!env || !out || !out_len) return -1;
    size_t len = tinyweb__envelope__get_packed_size((Tinyweb__Envelope*)env);
    unsigned char* buf = malloc(len);
    if (!buf) return -1;
    tinyweb__envelope__pack((Tinyweb__Envelope*)env, buf);
    *out = buf;
    *out_len = len;
    return 0;
}

Tinyweb__Envelope* tw_envelope_deserialize(const unsigned char* buf, size_t len) {
    if (!buf || !len) return NULL;
    return tinyweb__envelope__unpack(NULL, len, buf);
}

void tw_envelope_free(Tinyweb__Envelope* env) {
    if (env) tinyweb__envelope__free_unpacked(env, NULL);
}

int tw_envelope_build_and_sign(const tw_envelope_header_view_t* view,
                               const unsigned char* plaintext,
                               size_t plaintext_len,
                               Tinyweb__Envelope** out_env) {
    if (!view || !plaintext || !out_env) return -1;

    // Allocate envelope
    Tinyweb__Envelope* env = calloc(1, sizeof(Tinyweb__Envelope));
    if (!env) return -1;
    tinyweb__envelope__init(env);

    // Allocate and populate header
    Tinyweb__EnvelopeHeader* hdr = calloc(1, sizeof(Tinyweb__EnvelopeHeader));
    if (!hdr) {
        free(env);
        return -1;
    }
    tinyweb__envelope_header__init(hdr);
    
    hdr->version = view->version;
    hdr->content_type = view->content_type;
    hdr->schema_version = view->schema_version;
    hdr->timestamp = view->timestamp;
    
    // Allocate and copy sender pubkey
    hdr->sender_pubkey.len = PUBKEY_SIZE;
    hdr->sender_pubkey.data = malloc(PUBKEY_SIZE);
    if (!hdr->sender_pubkey.data) {
        free(hdr);
        free(env);
        return -1;
    }
    memcpy(hdr->sender_pubkey.data, view->sender_pubkey, PUBKEY_SIZE);

    // Allocate and copy recipients_pubkey
    hdr->n_recipients_pubkey = view->num_recipients;
    hdr->recipients_pubkey = calloc(view->num_recipients, sizeof(ProtobufCBinaryData));
    if (!hdr->recipients_pubkey) {
        free(hdr->sender_pubkey.data);
        free(hdr);
        free(env);
        return -1;
    }
    for (size_t i = 0; i < view->num_recipients; ++i) {
        hdr->recipients_pubkey[i].len = PUBKEY_SIZE;
        hdr->recipients_pubkey[i].data = malloc(PUBKEY_SIZE);
        if (!hdr->recipients_pubkey[i].data) {
            for (size_t j = 0; j < i; ++j) free(hdr->recipients_pubkey[j].data);
            free(hdr->recipients_pubkey);
            free(hdr->sender_pubkey.data);
            free(hdr);
            free(env);
            return -1;
        }
        memcpy(hdr->recipients_pubkey[i].data, view->recipients_pubkeys + i * PUBKEY_SIZE, PUBKEY_SIZE);
    }

    // Copy group_id if provided
    if (view->group_id && view->group_id_len > 0) {
        hdr->group_id.len = view->group_id_len;
        hdr->group_id.data = malloc(view->group_id_len);
        if (!hdr->group_id.data) {
            for (size_t j = 0; j < view->num_recipients; ++j) free(hdr->recipients_pubkey[j].data);
            free(hdr->recipients_pubkey);
            free(hdr->sender_pubkey.data);
            free(hdr);
            free(env);
            return -1;
        }
        memcpy(hdr->group_id.data, view->group_id, view->group_id_len);
    } else {
        hdr->group_id.data = NULL;
        hdr->group_id.len = 0;
    }

    env->header = hdr;

    // Encrypt plaintext directly into envelope
    if (encrypt_envelope_payload(plaintext, plaintext_len,
                                 view->recipients_pubkeys, view->num_recipients,
                                 env) != 0) {
        if (hdr->group_id.data) free(hdr->group_id.data);
        for (size_t j = 0; j < view->num_recipients; ++j) free(hdr->recipients_pubkey[j].data);
        free(hdr->recipients_pubkey);
        free(hdr->sender_pubkey.data);
        free(hdr);
        free(env);
        return -1;
    }

    // Compute digest and sign
    unsigned char digest[SHA256_DIGEST_LENGTH];
    if (compute_envelope_digest(env->header, env->payload_ciphertext.data, env->payload_ciphertext.len, digest) != 0) {
        tinyweb__envelope__free_unpacked(env, NULL);
        return -1;
    }

    env->signature.len = SIGNATURE_SIZE;
    env->signature.data = malloc(SIGNATURE_SIZE);
    if (!env->signature.data) {
        tinyweb__envelope__free_unpacked(env, NULL);
        return -1;
    }

    if (sign_message((const char*)digest, env->signature.data) != 0) {
        tinyweb__envelope__free_unpacked(env, NULL);
        return -1;
    }

    *out_env = env;
    return 0;
}

int tw_envelope_verify(const Tinyweb__Envelope* env) {
    if (!env) return -1;
    unsigned char digest[SHA256_DIGEST_LENGTH];
    if (compute_envelope_digest(env->header, env->payload_ciphertext.data, env->payload_ciphertext.len, digest) != 0) {
        logger_error("envelope", "Failed to compute envelope digest");
        return -1;
    }
    const unsigned char* pk = env->header->sender_pubkey.data;
    int result = verify_signature(env->signature.data, digest, sizeof(digest), pk);
    if (result != 0) {
        logger_error("envelope", "Signature verification failed: result=%d", result);
        logger_error("envelope", "Signature len=%zu, digest len=%zu, pubkey len=%zu", 
                    env->signature.len, sizeof(digest), env->header->sender_pubkey.len);
    }
    return result;
}

int tw_envelope_peek(const Tinyweb__Envelope* env,
                     const Tinyweb__EnvelopeHeader** out_hdr,
                     const unsigned char** out_cipher,
                     size_t* out_cipher_len) {
    if (!env || !out_hdr || !out_cipher || !out_cipher_len) return -1;
    *out_hdr = env->header;
    *out_cipher = env->payload_ciphertext.data;
    *out_cipher_len = env->payload_ciphertext.len;
    return 0;
}

File: client/web-ui/src/utils/envelope.js
import sodium from 'libsodium-wrappers';
import keyStore from './keystore.js';
import { encryptPayloadMulti, encryptedPayloadToHex } from './encryption.js';
import { loadProtobufSchema } from './protobufHelper.js';

/**
 * Frontend envelope utility - mirrors backend envelope logic
 * Handles message structure, encryption, and signing
 */

// Ensure sodium is ready before use
async function ensureSodiumReady() {
  await sodium.ready;
}

// Content types matching backend
export const CONTENT_TYPE = {
  DIRECT_MESSAGE: 1,
  GROUP_MESSAGE: 2,
  LOCATION_UPDATE: 3,
  EMERGENCY_ALERT: 4
};

// Envelope header structure
export class EnvelopeHeader {
  constructor() {
    this.version = 1;
    this.contentType = CONTENT_TYPE.DIRECT_MESSAGE;
    this.schemaVersion = 1;
    this.timestamp = Date.now();
    this.senderPubkey = null; // Will be set from keystore
    this.recipientPubkeys = [];
    this.groupId = null;
  }
}

/**
 * Create and sign an envelope with encrypted content
 * @param {EnvelopeHeader} header - The envelope header
 * @param {Uint8Array} plaintext - The message content
 * @returns {Object} - Signed envelope object
 */
export async function createSignedEnvelope(header, plaintext) {
  await ensureSodiumReady();
  
  if (!header.senderPubkey) {
    header.senderPubkey = keyStore.getPublicKey();
  }

  // Convert Ed25519 recipient pubkeys to X25519 for encryption
  const encryptionPubkeys = header.recipientPubkeys.map(ed25519Pubkey => 
    sodium.crypto_sign_ed25519_pk_to_curve25519(ed25519Pubkey)
  );

  // Encrypt the message for all recipients (using X25519 keys)
  const encrypted = await encryptPayloadMulti(plaintext, encryptionPubkeys);

  // Create envelope structure
  const envelope = {
    header: {
      version: header.version,
      contentType: header.contentType,
      schemaVersion: header.schemaVersion,
      timestamp: header.timestamp,
      senderPubkey: header.senderPubkey,
      recipientPubkeys: header.recipientPubkeys,
      groupId: header.groupId
    },
    encryptedPayload: encrypted,
    signature: null
  };

  // Compute digest and sign
  const digest = await computeEnvelopeDigest(envelope.header, encrypted.ciphertext);
  envelope.signature = await signDigest(digest);

  return envelope;
}

/**
 * Compute envelope digest for signing (matches backend logic exactly)
 * Uses raw header data (not protobuf-serialized) for cleaner digest computation
 * Backend uses: SHA256(domain || raw_header_fields || SHA256(ciphertext))
 * @param {Object} header - Envelope header
 * @param {Uint8Array} ciphertext - Encrypted content
 * @returns {Promise<Uint8Array>} - SHA256 digest (32 bytes)
 */
async function computeEnvelopeDigest(header, ciphertext) {
  await ensureSodiumReady();
  
  // Create domain separator with null terminator (matches backend: "TWENVELOPE\0")
  const domain = new Uint8Array([84, 87, 69, 78, 86, 69, 76, 79, 80, 69, 0]); // "TWENVELOPE\0"

  // Hash ciphertext with SHA256 first
  const payloadHash = await sha256(ciphertext);

  // Build digest from raw header fields (canonical order matching backend)
  const parts = [];
  
  // 1. Domain
  parts.push(domain);
  
  // 2. version (uint32, 4 bytes, little-endian)
  const versionBytes = new Uint8Array(4);
  const versionView = new DataView(versionBytes.buffer);
  versionView.setUint32(0, header.version, true); // little-endian
  parts.push(versionBytes);
  
  // 3. content_type (uint32, 4 bytes, little-endian)
  const contentTypeBytes = new Uint8Array(4);
  const contentTypeView = new DataView(contentTypeBytes.buffer);
  contentTypeView.setUint32(0, header.contentType, true);
  parts.push(contentTypeBytes);
  
  // 4. schema_version (uint32, 4 bytes, little-endian)
  const schemaVersionBytes = new Uint8Array(4);
  const schemaVersionView = new DataView(schemaVersionBytes.buffer);
  schemaVersionView.setUint32(0, header.schemaVersion, true);
  parts.push(schemaVersionBytes);
  
  // 5. timestamp (uint64, 8 bytes, little-endian) - convert ms to seconds
  const timestampBytes = new Uint8Array(8);
  const timestampView = new DataView(timestampBytes.buffer);
  const timestampSeconds = Math.floor(header.timestamp / 1000);
  // Use setUint32 for 64-bit value (split into two 32-bit parts for compatibility)
  // JavaScript numbers are safe up to 2^53, so we can use this approach
  const low = timestampSeconds & 0xFFFFFFFF;
  const high = Math.floor(timestampSeconds / 0x100000000);
  timestampView.setUint32(0, low, true);  // little-endian
  timestampView.setUint32(4, high, true); // little-endian
  parts.push(timestampBytes);
  
  // 6. sender_pubkey (32 bytes)
  const senderPubkey = header.senderPubkey instanceof Uint8Array 
    ? header.senderPubkey 
    : new Uint8Array(header.senderPubkey);
  if (senderPubkey.length !== 32) {
    throw new Error(`Invalid sender_pubkey length: ${senderPubkey.length}, expected 32`);
  }
  parts.push(senderPubkey);
  
  // 7. recipients_pubkey (count as uint32, then each 32-byte pubkey)
  const numRecipientsBytes = new Uint8Array(4);
  const numRecipientsView = new DataView(numRecipientsBytes.buffer);
  numRecipientsView.setUint32(0, header.recipientPubkeys.length, true);
  parts.push(numRecipientsBytes);
  
  for (const recipientPubkey of header.recipientPubkeys) {
    const pk = recipientPubkey instanceof Uint8Array 
      ? recipientPubkey 
      : new Uint8Array(recipientPubkey);
    if (pk.length !== 32) {
      throw new Error(`Invalid recipient_pubkey length: ${pk.length}, expected 32`);
    }
    parts.push(pk);
  }
  
  // 8. group_id (length as uint32, then data, or empty)
  const groupId = header.groupId 
    ? (header.groupId instanceof Uint8Array ? header.groupId : new Uint8Array(header.groupId))
    : new Uint8Array(0);
  const groupIdLenBytes = new Uint8Array(4);
  const groupIdLenView = new DataView(groupIdLenBytes.buffer);
  groupIdLenView.setUint32(0, groupId.length, true);
  parts.push(groupIdLenBytes);
  if (groupId.length > 0) {
    parts.push(groupId);
  }
  
  // 9. Payload hash
  parts.push(payloadHash);
  
  // Combine all parts and hash
  const totalLength = parts.reduce((sum, part) => sum + part.length, 0);
  const combined = new Uint8Array(totalLength);
  let offset = 0;
  for (const part of parts) {
    combined.set(part, offset);
    offset += part.length;
  }
  
  return await sha256(combined);
}

/**
 * Compute SHA256 hash (using Web Crypto API for compatibility)
 * @param {Uint8Array} data - Data to hash
 * @returns {Promise<Uint8Array>} - SHA256 hash (32 bytes)
 */
async function sha256(data) {
  const hashBuffer = await crypto.subtle.digest('SHA-256', data);
  return new Uint8Array(hashBuffer);
}

/**
 * Sign a digest using the keystore
 * @param {Uint8Array} digest - The digest to sign
 * @returns {Uint8Array} - The signature
 */
async function signDigest(digest) {
  await ensureSodiumReady();
  const privateKey = keyStore._getPrivateKey();
  return sodium.crypto_sign_detached(digest, privateKey);
}

/**
 * Verify an envelope signature
 * @param {Object} envelope - The envelope to verify
 * @returns {boolean} - Whether signature is valid
 */
export async function verifyEnvelopeSignature(envelope) {
  await ensureSodiumReady();
  
  console.log('🔍 verifyEnvelopeSignature called');
  console.log('📦 Envelope:', envelope);
  console.log('📋 Header:', envelope.header);
  console.log('👤 Sender pubkey:', envelope.header.senderPubkey);
  console.log('👤 Sender pubkey type:', envelope.header.senderPubkey?.constructor?.name);
  console.log('👤 Sender pubkey length:', envelope.header.senderPubkey?.length);
  console.log('🔑 Signature:', envelope.signature);
  console.log('🔑 Signature type:', envelope.signature?.constructor?.name);
  console.log('🔑 Signature length:', envelope.signature?.length);
  
  const digest = await computeEnvelopeDigest(envelope.header, envelope.encryptedPayload.ciphertext);
  console.log('🔐 Computed digest:', digest);
  console.log('🔐 Digest length:', digest?.length);
  
  try {
    const result = sodium.crypto_sign_verify_detached(
      envelope.signature,
      digest,
      envelope.header.senderPubkey
    );
    console.log('✅ Verification result:', result);
    return result;
  } catch (err) {
    console.error('❌ crypto_sign_verify_detached error:', err);
    throw err;
  }
}

/**
 * Convert envelope to transaction format for API submission
 * @param {Object} envelope - The signed envelope
 * @returns {Object} - Transaction data ready for API
 */
export async function envelopeToTransaction(envelope) {
  await ensureSodiumReady();
  
  // Serialize the envelope (simplified - in production would use protobuf)
  const envelopeData = {
    header: envelope.header,
    encryptedPayloadHex: await encryptedPayloadToHex(envelope.encryptedPayload),
    signatureHex: sodium.to_hex(envelope.signature)
  };

  // Convert to the format expected by the backend API
  const transaction = {
    type: getTransactionTypeForContent(envelope.header.contentType),
    sender: sodium.to_hex(envelope.header.senderPubkey),
    timestamp: envelope.header.timestamp,
    recipients: envelope.header.recipientPubkeys.map(pk => sodium.to_hex(pk)),
    recipientCount: envelope.header.recipientPubkeys.length,
    payload: envelopeData, // This would be serialized properly in production
    signature: envelope.signature
  };

  return transaction;
}

/**
 * Convert content type to transaction type
 * @param {number} contentType - Envelope content type
 * @returns {number} - Transaction type
 */
function getTransactionTypeForContent(contentType) {
  switch (contentType) {
    case CONTENT_TYPE.DIRECT_MESSAGE:
    case CONTENT_TYPE.GROUP_MESSAGE:
      return 23; // TW_TXN_MESSAGE or TW_TXN_GROUP_MESSAGE
    case CONTENT_TYPE.LOCATION_UPDATE:
      return 37; // TW_TXN_LOCATION_UPDATE
    case CONTENT_TYPE.EMERGENCY_ALERT:
      return 38; // TW_TXN_EMERGENCY_ALERT
    default:
      return 23; // Default to message
  }
}

/**
 * Create a direct message envelope
 * @param {Uint8Array} recipientPubkey - Recipient's public key
 * @param {string} message - The message text
 * @returns {Object} - Signed envelope
 */
export async function createDirectMessage(recipientPubkey, message) {
  const header = new EnvelopeHeader();
  header.contentType = CONTENT_TYPE.DIRECT_MESSAGE;
  header.recipientPubkeys = [recipientPubkey];

  const plaintext = new TextEncoder().encode(message);
  return await createSignedEnvelope(header, plaintext);
}

/**
 * Create a group message envelope
 * @param {Uint8Array[]} recipientPubkeys - Array of recipient public keys
 * @param {string} message - The message text
 * @param {Uint8Array} groupId - Optional group identifier
 * @returns {Object} - Signed envelope
 */
export async function createGroupMessage(recipientPubkeys, message, groupId = null) {
  const header = new EnvelopeHeader();
  header.contentType = CONTENT_TYPE.GROUP_MESSAGE;
  header.recipientPubkeys = recipientPubkeys;
  header.groupId = groupId;

  const plaintext = new TextEncoder().encode(message);
  return await createSignedEnvelope(header, plaintext);
}
